src/audits/DuplicateId.js

```diff
@@ -25,15 +25,36 @@ axs.AuditRules.addRule({
     heading: 'Any ID referred to via an IDREF must be unique in the DOM',
     url: 'https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_html_02',
     severity: axs.constants.Severity.SEVERE,
-    relevantElementMatcher: function(element) {
-        if (element.hasAttribute('id')) {
-            var referrers = axs.utils.getIdReferrers(element);
-            return referrers.some(function(referrer) {
-                return !axs.utils.isElementOrAncestorHidden(referrer);
+    opt_requires: {
+        domWalk: true,
+        idRefs: true
+    },
+    /**
+     * @this {axs.AuditRule}
+     */
+    relevantElementMatcher: function(element, flags) {
+        if (flags.idrefs.length && !flags.hidden) {
+            this.relatedElements.push({
+                element: element,
+                flags: flags
             });
         }
+        if (element.hasAttribute('id')) {
+            return true;
+        }
         return false;
     },
+    /**
+     * @this {axs.AuditRule}
+     */
+    isRelevant: function(element, flags) {
```

Comment:
Yep it's hard to follow, I should have commented it more and I'm glad you're looking at this while it's still in my head.

I can split it up, sure thing.
If you're in my neighbourhood feel free to come to my place and we could go over it together. It really does need explaining.

This particular line `isRelevant` is necessary to faithfully preserve the `DuplicateId` audit's behaviour:

- It needs to return `NA` if there are no relevant elements.
- Relevant elements are those which have one or more "id referrers".

The problem: ID referrers may be encountered **after** the element, so we have to wait until the DOM traversal is totally finished before we can know if there are *really* relevant elements.

Without `isRelevant` DuplicateId would return `PASS` instead of `NA` for every element with an ID.